Dynamic EBS Volume Sizing

herbertgoto opened the pull request:
Issue #:
kubernetes-sigs/karpenter#1988

Description of changes:
This blueprint shows how to automatically resize EBS volumes based on the EC2 instance type that Karpenter provisions. EBS volume size requirements differ among different instance types and this pattern ensures that each node gets an appropriately sized root volume without manual intervention.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choic

herbertgoto committed:
feat(karpenter): add dynamic disk karpenter assets - fix 8

herbertgoto committed:
feat(karpenter): add dynamic disk karpenter assets - fix 9

herbertgoto committed:
feat(karpenter): add dynamic disk karpenter assets - fix 12

herbertgoto committed:
feat(karpenter): add dynamic disk karpenter assets

herbertgoto committed:
Merge branch 'aws-samples:main' into main

herbertgoto committed:
feat(karpenter): add dynamic disk karpenter for Bottlerocket

herbertgoto committed:
Merge branch 'main' of github.com:herbertgoto/karpenter-blueprints

herbertgoto committed:
feat(karpenter): add dynamic disk karpenter for Bottlerocket - README update

chrismld reviewed (changes requested):
hello Herbert, thanks for your contribution!! I left a few comments. For me, the blueprint didn't work, would you mind checking again? let me know if you'd like to troubleshoot with my setup :)

herbertgoto committed:
Merge branch 'aws-samples:main' into main

herbertgoto committed:
feat(karpenter): resolve review recommendations

herbertgoto commented on blueprints/dynamic-disk-ebs-volume/README.md:
````diff
@@ -0,0 +1,92 @@
+And the workload pods are now running:
+
+```sh
+> kubectl get pods                                                                                                             
````
It normally takes around 1 mins.

chrismld reviewed (changes requested):
thanks again Herb, the blueprint works for me now ... still, I'm requesting a few more changes to make a few things even easier to understand. Also, there's one important change that we need in order to test the whole blueprint easily. Also, please don't solve the conversation as it helps me to track what my feedback was :P

herbertgoto commented on blueprints/dynamic-disk-ebs-volume/workload.yaml:
```diff
@@ -0,0 +1,23 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: dynamic-disk-ebs-volume
```
I can duplicate the instructions in each section but it makes no sense to duplicate the deployment manifest since it is the same.

herbertgoto commented on blueprints/dynamic-disk-ebs-volume/README.md:
```diff
@@ -0,0 +1,166 @@
+
+</details>
+
+Once you have deployed the `EC2NodeClass` and `Nodepool` for Amazon Linux 2023 or Bottlecket, proceed to deploy the test workload:
```
This is a test workload that it is intended to show that the volume was actually resized. 
Is it really relevant that the test complies with the use case?

herbertgoto committed:
feat(karpenter): PR adjustments

herbertgoto commented on blueprints/dynamic-disk-ebs-volume/al2023.yaml:
```diff
@@ -0,0 +1,251 @@
+        volumeType: gp3
+        deleteOnTermination: true
+        encrypted: true
+  userData: |
```
I consolidated everything in one script but it has to be pass by differently to each OS. 
The steps are the same in the guide.

chrismld reviewed (changes requested):
thanks for the new changes Herb, there are a few things that still need to be resolved and I added a comment

herbertgoto committed:
feat(karpenter): refactor the blueprint proposal

herbertgoto commented on blueprints/dynamic-disk-ebs-volume/al2023.yaml:
```diff
@@ -0,0 +1,251 @@
+apiVersion: karpenter.k8s.aws/v1
+kind: EC2NodeClass
+metadata:
+  name: dynamic-disk-volume
```
Fixed. I just left the NodePool for AL2023.

chrismld commented on blueprints/dynamic-disk-ebs-volume/ebs-dynamic-resize.yaml:
```diff
@@ -0,0 +1,239 @@
+            echo "al2023"
+        fi
+    }
+
```
@erezzarum do you have any comment about this?

chrismld reviewed (commented):
LGTM Herb, I'll just wait to hear back from @erezzarum about the Bottlerocket question in case we need to add an update or clarification to the main README file. Thanks again!!!

herbertgoto committed:
feat(karpenter): add changes for NVMe ephemeral storage instances

herbertgoto commented:
@chrismld @pawelros I ran several tests and made changes to support instances with ephemeral storage:
- I added `instanceStorePolicy: RAID0` to the NodePool which sets up a RAID-0 XFS filesystem from any NVMe instance storage disks, moves the contents of `/var/lib/kubelet`, `/var/log/pods`, and `/var/lib/containerd` to the new RAID, and symlinks those directories back to the root filesystem.
- I modified the bootstrap script to not resize EBS volumes for instances with ephemeral storage. 

Following the results for some tests:
```
AL2023 - instance with NVMe

NAME          MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
nvme0n1       259:0    0    20G  0 disk
├─nvme0n1p1   259:2    0    20G  0 part  /
├─nvme0n1p127 259:3    0     1M  0 part
└─nvme0n1p128 259:4    0    10M  0 part  /boot/efi
nvme1n1       259:1    0 109.9G  0 disk
└─md127         9:127  0 109.8G  0 raid0 /var/log/pods
                                         /var/lib/containerd
                                         /var/lib/kubelet
                                         /mnt/k8s-disks/0

AL2023 - instance without NVMe

sh-5.2$ lsblk
NAME          MAJ:MIN RM SIZE RO TYPE MOUNTPOINTS
nvme0n1       259:0    0  60G  0 disk
├─nvme0n1p1   259:1    0  60G  0 part /
├─nvme0n1p127 259:2    0   1M  0 part
└─nvme0n1p128 259:3    0  10M  0 part /boot/efi

BR - instance with NVMe 

NAME         MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
nvme0n1      259:0    0     4G  0 disk
|-nvme0n1p1  259:3    0     4M  0 part
|-nvme0n1p2  259:4    0     5M  0 part
|-nvme0n1p3  259:5    0    40M  0 part /boot
|-nvme0n1p4  259:7    0   920M  0 part
|-nvme0n1p5  259:8    0    10M  0 part
|-nvme0n1p6  259:9    0    25M  0 part
|-nvme0n1p7  259:10   0     5M  0 part
|-nvme0n1p8  259:11   0    40M  0 part
|-nvme0n1p9  259:12   0   920M  0 part
|-nvme0n1p10 259:13   0    10M  0 part
|-nvme0n1p11 259:14   0    25M  0 part
|-nvme0n1p12 259:15   0    41M  0 part /var/lib/bottlerocket
|                                      /.bottlerocket
`-nvme0n1p13 259:16   0     1M  0 part
nvme2n1      259:1    0 109.9G  0 disk /var/log/pods
                                       /var/lib/kubelet
                                       /var/lib/containerd
                                       /mnt/.ephemeral
nvme1n1      259:2    0    20G  0 disk
`-nvme1n1p1  259:17   0    20G  0 part /var
                                       /opt
                                       /mnt
                                       /local

BR - instance without NVMe 

NAME         MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
nvme1n1      259:0    0   60G  0 disk
`-nvme1n1p1  259:12   0   60G  0 part /opt
                                      /var
                                      /mnt
                                      /local
nvme0n1      259:1    0    4G  0 disk
|-nvme0n1p1  259:2    0    4M  0 part
|-nvme0n1p2  259:3    0    5M  0 part
|-nvme0n1p3  259:4    0   40M  0 part /boot
|-nvme0n1p4  259:5    0  920M  0 part
|-nvme0n1p5  259:6    0   10M  0 part
|-nvme0n1p6  259:7    0   25M  0 part
|-nvme0n1p7  259:8    0    5M  0 part
|-nvme0n1p8  259:9    0   40M  0 part
|-nvme0n1p9  259:10   0  920M  0 part
|-nvme0n1p10 259:11   0   10M  0 part
|-nvme0n1p11 259:13   0   25M  0 part
|-nvme0n1p12 259:14   0   41M  0 part /var/lib/bottlerocket
|                                     /.bottlerocket
`-nvme0n1p13 259:15   0    1M  0 part
```

chrismld reviewed (approved):
Thanks @herbertgoto 🚀 ... this LGTM!!

mark-webster-catalyst commented:
I cannot for the life of me get this to work with bottlerocket. I can see the EBS volumes in the AWS console that are the correct (resized) size, but the node doesn't recognise the extra capacity. Any pointers?

herbertgoto commented:
hey @mark-webster-catalyst

I copy the bash script in a sh file and base64 it:
`cat base.sh | base64 -w 0 | pbcopy`

Then I created this ec2nodeclass and nodepool, I replaced the base64 script in the userData:

```yaml
apiVersion: karpenter.k8s.aws/v1
kind: EC2NodeClass
metadata:
  name: ebs-dynamic-resize
spec:
  amiSelectorTerms:
  - alias: bottlerocket@v1.57.0
  role: "karpenter-blueprints"
  securityGroupSelectorTerms:
  - tags:
      karpenter.sh/discovery: "karpenter-blueprints"
  subnetSelectorTerms:
  - tags:
      karpenter.sh/discovery: "karpenter-blueprints"
  metadataOptions:
    httpEndpoint: enabled
    httpTokens: required
    httpPutResponseHopLimit: 1
  instanceStorePolicy: RAID0
  blockDeviceMappings:
    # Root device
    - deviceName: /dev/xvda
      ebs:
        volumeSize: 4Gi
        volumeType: gp3
        encrypted: true
    # Data device: Container resources such as images and logs
    - deviceName: /dev/xvdb
      ebs:
        volumeSize: 20Gi
        volumeType: gp3
        encrypted: true
  userData: |
    [settings.bootstrap-containers.ebsresize]
    mode = "once"
    user-data = "BASE64_USER_DATA"
---
apiVersion: karpenter.sh/v1
kind: NodePool
metadata:
  name: ebs-dynamic-resize
spec:
  disruption:
    consolidationPolicy: WhenEmptyOrUnderutilized
    consolidateAfter: 1m
  limits:
    cpu: 1k
    memory: 500Gi
  template:
    metadata:
      labels:
        intent: ebs-dynamic-resize
    spec:
      nodeClassRef:
        group: karpenter.k8s.aws
        kind: EC2NodeClass
        name: ebs-dynamic-resize
      requirements:
      - key: karpenter.sh/capacity-type
        operator: In
        values:
        - spot
        - on-demand
      - key: kubernetes.io/os
        operator: In
        values:
        - linux
      - key: kubernetes.io/arch
        operator: In
        values:
        - amd64
      - key: karpenter.k8s.aws/instance-category
        operator: In
        values:
        - c
        - m
        - r
```
It worked for me!
I had EKS 1.34 and public.ecr.aws/karpenter/controller:1.8.3

herbertgoto commented:
BTW, I tested without the property you mentioned in this [issue](https://github.com/aws-samples/karpenter-blueprints/issues/48).
